linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/ConstantArrayPool.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.backend.wasmemitter
+
+import java.nio.{ByteBuffer, ByteOrder}
+
+import scala.collection.mutable
+
+import org.scalajs.ir.OriginalName
+
+import org.scalajs.linker.backend.wasmemitter.VarGen.genDataID
+
+import org.scalajs.linker.backend.webassembly.Identitities._
+import org.scalajs.linker.backend.webassembly.Modules._
+
+/** Pool of constant arrays that we store in data segments. */
+final class ConstantArrayPool {
+  /* We use 4 data segments; one for each byte size: 1, 2, 4 and 8.
+   * This way, every sub-segment containing the contents of an array is aligned
+   * to the byte size of elements of that array.
+   */
+
+  // Indexed by log2ByteSize
+  private val constantArrays = Array.fill(4)(mutable.ListBuffer.empty[Array[Byte]])
```

Comment:
I assume you have considered using an array buffer and building the entire data right away?